Add prepare_database function in prepdb.py

murphyke opened the pull request:
Fix #3

aaron0browne commented on pedsnetdcc/prepdb.py:
```diff
@@ -0,0 +1,240 @@
+logger = logging.getLogger(__name__)
+
+
+def _conn_str_with_database(conn_str, dbname):
```
I'm guessing these first four functions will move to `utils.py`, but we can wait until they are actually needed by other modules so as not to be moving things back and forth needlessly.

aaron0browne commented on pedsnetdcc/prepdb.py:
```diff
@@ -0,0 +1,240 @@
+create schema if not exists {{.Site}}_pedsnet authorization dcc_owner;
+create schema if not exists {{.Site}}_pcornet authorization dcc_owner;
+create schema if not exists {{.Site}}_harvest authorization dcc_owner;
+create schema if not exists {{.Site}}_achilles authorization dcc_owner;
```
Please add `create schema if not exists {{.Site}}_id_maps authorization dcc_owner;` for the id mapping.

murphyke replied:
ok

aaron0browne commented on pedsnetdcc/prepdb.py:
```diff
@@ -0,0 +1,240 @@
+grant usage  on               schema {{.Site}}_pcornet  to peds_staff;
+grant select on all tables in schema {{.Site}}_pcornet  to peds_staff;
+grant all    on               schema {{.Site}}_harvest  to harvest_user;
+grant all    on               schema {{.Site}}_achilles to achilles_user;
```
Please add `grant all on schema {{.Site}}_id_maps to loading_user;` for the id mapping.

murphyke replied:
ok

murphyke replied:
Should we add default select privilege for peds_staff?

aaron0browne replied:
Good idea.

aaron0browne commented on pedsnetdcc/prepdb.py:
```diff
@@ -0,0 +1,240 @@
+"""
+
+# SQL template for creating site schemas in a production database instance.
+_sql_site_prod_template = """
```
We haven't talked yet, that I remember, about whether the id maps get moved to prod, but it seems to me like they probably should, in which case the above two statements should also be added here.

murphyke replied:
Sounds reasonable to me also

aaron0browne commented on pedsnetdcc/prepdb.py:
```diff
@@ -0,0 +1,244 @@
+    return [x for x in sql.split("\n") if x]
+
+
+_sql_vocab_internal = """
```
We also need a `dcc_ids` schema to hold the `last_id` value for each domain during the id mapping (see [here](https://github.com/PEDSnet/dcc-loader/blob/master/id-map/create-all-schemas.sql#L135)). The pertinent statements are:

```
create schema if not exists dcc_ids authorization dcc_owner;
grant all on schema dcc_ids to loading_user;
```

I don't think anyone else besides the `loading_user` will need access to this schema. We could shove it into the vocab sql lists, since that only gets executed once, or make a separate set for it. Up to you.

aaron0browne commented on pedsnetdcc/prepdb.py:
```diff
@@ -0,0 +1,244 @@
+_sql_vocab_prod = """
+grant usage                on schema vocabulary to harvest_user, peds_staff;
+grant select on all tables in schema vocabulary to harvest_user, peds_staff;
+alter default privileges for role dcc_owner in schema vocabulary grant select on tables to harvest_user, peds_staff;
```
`s/dcc_owner/loading_user/`

aaron0browne replied:
Oh, wait, maybe I don't understand. You're expecting the `dcc_owner` role to be used for moving data from the internal db to the prod db, is that right?

murphyke replied:
That was the original idea, that `loading_user` wouldn't even exist in prod -- only the users actually necessary.  However, I think at some point I swung around and decided that it would be easier to just have all users created in both databases. Then we can do vanilla dumps and restores and have just one source of users to maintain, and hopefuly one set of privileges.

aaron0browne commented on pedsnetdcc/prepdb.py:
```diff
@@ -0,0 +1,244 @@
+    logger.info({'msg': 'Starting database preparation.'})
+
+    with psycopg2.connect(conn_str) as conn:
+        conn.set_isolation_level(0)
```
If there were an `isolation_level` argument to the `db.py` functions, this function could use those and not need to worry about error handling or messaging properly, right?

murphyke replied:
Yes, I vote for that and can rewrite to use the Statement* API.

aaron0browne commented on pedsnetdcc/prepdb.py:
```diff
@@ -0,0 +1,244 @@
+    :param update: assume the database is already created
+    :type: bool
+    :param dcc_only: only create schemas for `dcc` (no sites)
+    :return: None, or (if `psql` is True) a SQL string
```
Seems like the `psql` related string only functionality is gone now.

murphyke replied:
Thanks.

aaron0browne commented on pedsnetdcc/tests/prepdb_test.py:
```diff
@@ -0,0 +1,239 @@
+             'pcor_et_user', 'loading_user', 'dcc_owner', 'pcornet_sas')
+
+    def create_roles(self):
+        """Create roles so grants will not fail.
```
This makes me wonder why role creation isn't part of the `prepare_database` function?

murphyke replied:
Yes, I thought about the same thing and left it this way just because the dbhi/pedsnet scripts had already staked out that turf. But it would be best to gather everything under `pedsnetdcc`, in which case, either `prepdb` could subsume the role creation and handle errors when roles already exist, or it could invoke a role creation method in a possible `createcluster` module.

aaron0browne replied:
The role creation steps still need to be added, right?

murphyke replied:
As per convo, the roles are independent of individual databases and are the purview of the internal dbhi/pedsnet repo. Until that repo is subsumed into this tool, we will leave this as is.

aaron0browne commented on pedsnetdcc/tests/prepdb_test.py:
```diff
@@ -0,0 +1,239 @@
+        conn.close()
+        return True
+
+    def setUp(self):
```
Probably worth changing and simplifying a lot of this using the `testing.postgresql` package.

murphyke commented:
Addressed review comments. Biggest changes:

* Alter db.py to support non-transactional statements like CREATE DATABASE.
* Update prepdb to use db.py API.
* Abandon distinction between prod and internal -- internal is the new default.
* id maps stuff added

aaron0browne commented on pedsnetdcc/tests/db_test.py:
```diff
@@ -132,6 +140,37 @@ def test_serial_execute_transaction(self):
+        stmts.append(Statement('VACUUM'))
+        stmts.serial_execute(self.conn_str, True)
+
+        self.assertIn('VACUUM cannot run inside a transaction block',
```
I believe that `assertIn` was added in python 3.2, so this should be `self.assertTrue(foo in bar)` for python 2 compatibility.

murphyke replied:
I see that in the 2.X line, `assertIn` was added in 2.7. Can we live with that?

aaron0browne replied:
Oh I was just looking at the 3 docs where it says added in 3.1. Confusing. Yes, 2.7 is as far back as we are going. So ignore this comment and the next one about `assertIsNone`.

aaron0browne commented on pedsnetdcc/tests/db_test.py:
```diff
@@ -132,6 +140,37 @@ def test_serial_execute_transaction(self):
+        stmts.append(Statement('VACUUM'))
+        stmts.serial_execute(self.conn_str)
+
+        self.assertIsNone(stmts[0].err)
```
`self.assertTrue(foo is None)` for python 2 compatibility.

aaron0browne commented on pedsnetdcc/prepdb.py:
```diff
@@ -0,0 +1,212 @@
+    stmts.serial_execute(conn_str)
+
+    for stmt in stmts:
+        if stmt.err:
```
You could use `check_stmt_err` here, instead of the error tracking you are doing. That function both logs with the full `err` from the `Statement` and also raises an error which can be handled by the calling function (in this case the CLI). And it takes an argument so that you can describe what process was running, like `"database preparation"`.

murphyke replied:
My take was that the error had already been logged, so it was enough to return under control with an error status.

aaron0browne commented on pedsnetdcc/prepdb.py:
```diff
@@ -0,0 +1,212 @@
+    :param dcc_only: only create schemas for `dcc` (no sites)
+    :return: True on success, False otherwise
+    """
+    logger.info({'msg': 'Starting prepdb for {0}.'.format(model_version)})
```
`s/prepdb/database preparation/` for consistency with ending message. May be worth setting the start time and reporting the elapsed time at the end? Lower case and no period for merging with other error messages. The `model_version` should be a separate key/value in the dict, like: `{'msg': 'starting database preparation', 'model_version': model_version'}`.

aaron0browne commented on pedsnetdcc/tests/prepdb_test.py:
```diff
@@ -0,0 +1,182 @@
+        """Test for sites + dcc"""
+        prepare_database(self.model_version, self.conn_str)
+        self.check()
+        self.drop_database()
```
Is this still needed now that `testing.postgresql` is handling the dbs?

murphyke replied:
No longer needed; I thought testing.postgresql was caching the cluster, but it is caching the pre-cluster initted files; returns a fresh cluster each time.

murphyke commented:
Latest tweaks.

aaron0browne commented:
This looks great. Merging.
